packages/@lwc/template-compiler/src/parser/parser.ts

```diff
@@ -0,0 +1,159 @@
+    }
+
+    withErrorRecovery<T>(fn: () => T): T | undefined {
+        try {
+            return fn();
+        } catch (error) {
+            if (error instanceof CompilerError) {
+                this.addDiagnostic(error.toDiagnostic());
+            } else {
+                throw error;
+            }
+        }
+    }
+
+    withErrorWrapping<T>(fn: () => T, errorInfo: LWCErrorInfo, location: parse5.Location): T {
+        try {
+            return fn();
+        } catch (error) {
+            this.throwOnError(errorInfo, error, location);
+        }
+    }
+
+    warnOnIRNode(
+        errorInfo: LWCErrorInfo,
+        irNode: IRNode | IRBaseAttribute,
+        messageArgs?: any[]
+    ): void {
+        this.warnAtLocation(errorInfo, messageArgs, irNode.location);
+    }
+
+    warnAtLocation(errorInfo: LWCErrorInfo, messageArgs?: any[], location?: parse5.Location): void {
+        this.addDiagnostic(
+            generateCompilerDiagnostic(errorInfo, {
+                messageArgs,
+                origin: {
+                    location: normalizeLocation(location),
+                },
+            })
+        );
+    }
+
+    throwOnError(errorInfo: LWCErrorInfo, error: any, location?: parse5.Location): never {
+        const diagnostic = normalizeToDiagnostic(errorInfo, error, {
+            location: normalizeLocation(location),
+        });
+        throw CompilerError.from(diagnostic);
+    }
+
+    throwOnIRNode(
+        errorInfo: LWCErrorInfo,
+        irNode: IRNode | IRBaseAttribute,
+        messageArgs?: any[]
+    ): never {
+        this.throwAtLocation(errorInfo, irNode.location, messageArgs);
+    }
+
+    throwAtLocation(
+        errorInfo: LWCErrorInfo,
+        location?: parse5.Location,
+        messageArgs?: any[]
+    ): never {
+        throw generateCompilerError(errorInfo, {
+            messageArgs,
+            origin: {
+                location: normalizeLocation(location),
+            },
+        });
+    }
+
+    addDiagnostic(diagnostic: CompilerDiagnostic): void {
+        this.warnings.push(diagnostic);
+    }
+
+    getRoot(element: IRElement): IRElement {
+        return this.parentStack[0] || element;
+    }
+
+    getRenderMode(element: IRElement): LWCDirectiveRenderMode {
+        return this.getRoot(element).lwc?.renderMode ?? LWCDirectiveRenderMode.shadow;
+    }
+
+    getPreserveComments(element: IRElement): boolean {
+        return (
+            this.getRoot(element).lwc?.preserveComments?.value ?? this.config.preserveHtmlComments
+        );
+    }
+
+    isInIteration(element: IRElement): boolean {
+        let current: IRElement | undefined = element;
+
+        for (let i = this.parentStack.length; i >= 0; i--) {
+            if (current.tag === 'template') {
+                if (current.forEach || current.forOf) {
+                    return true;
+                }
+            }
+
+            current = this.parentStack[i - 1];
+        }
+
+        return false;
+    }
```

Comment:
I am not a big fan of having this method in the `ParserCtx`. We also have 2 other methods in `expression.ts` that are also using a similar pattern by climbing the parent stack. Those methods can be abstracted away by exposing a helper method like:

```ts
// Proposed signature
class ParserCtx {
  findAncestor(el: IRElement, predicate: (elm: IRElement) => unkown): IRElement
}

// Usage
const iterationParent = ctx.findAncestor(elm => elm.forEach || elm.forOff);
```

We can defer this in a follow-up PR, as I really want to get this merged.

Reply:
I'll add this one to the next PR as well, I figured since we had getRoot, getRenderMode and getPreserveComments we could put it in here for now.  

I'll move it back since it will be replaced once we have the new findAncestor API.